src/include/processor/operator/recursive_extend/bfs_state.h

```diff
@@ -30,18 +83,85 @@ class TargetDstNodes {
 
     inline uint64_t getNumNodes() const { return numNodes; }
 
+    inline frontier::node_id_set_t getNodeIDs() const { return nodeIDs; }
+
 private:
     uint64_t numNodes;
     frontier::node_id_set_t nodeIDs;
     std::unordered_set<common::table_id_t> tableIDFilter;
 };
 
-class BaseBFSState {
+struct BaseBFSMorsel;
+
+/**
+ * An SSSPSharedState is a unit of work for the nTkS scheduling policy. It is *ONLY* used for the
+ * particular case of SHORTEST_PATH | SINGLE_LABEL | NO_PATH_TRACKING. An SSSPSharedState is *NOT*
+ * exclusive to a thread and any thread can pick up BFS extension or Writing Path Length.
+ * A shared_ptr is maintained by the BaseBFSMorsel and a morsel of work is fetched using this ptr.
+ */
+struct SSSPSharedState {
+public:
+    SSSPSharedState(uint64_t upperBound_, uint64_t lowerBound_, uint64_t maxNodeOffset_)
+        : mutex{std::mutex()}, ssspLocalState{EXTEND_IN_PROGRESS}, currentLevel{0u},
+          nextScanStartIdx{0u}, numVisitedNodes{0u}, visitedNodes{std::vector<uint8_t>(
+                                                         maxNodeOffset_ + 1, NOT_VISITED)},
+          pathLength{std::vector<uint8_t>(maxNodeOffset_ + 1, 0u)},
+          bfsLevelNodeOffsets{std::vector<common::offset_t>()}, srcOffset{0u},
+          maxOffset{maxNodeOffset_}, upperBound{upperBound_}, lowerBound{lowerBound_},
+          numThreadsActiveOnMorsel{0u}, nextDstScanStartIdx{0u}, inputFTableTupleIdx{0u} {}
+
+    void reset(TargetDstNodes* targetDstNodes);
+
+    bool getBFSMorsel(std::unique_ptr<BaseBFSMorsel>& bfsMorsel, SSSPLocalState& state);
```

Comment:
Instead of passing `SSSPLocalState&` why not just return state. Also just give `BaseBFSMorsel*` as parameter change across PR.

Reply:
changed